### PR TITLE
TypeExtension.GetReadableName Improvements

### PR DESCRIPTION
**Pull request by mbaker3**
Improved `TypeExtension.GetReadableName` to:
 - Include outer types in the names of inner types.
 - Correctly handle types with a generic outer type - `MyType<int>.MyInnerType`

## TODO
Once this is approved there is some work to do copying the changes.
 - [x] ~Copy `GetReadableName()` logic into the Logging DLL~

## Tag Alongs
 - `MathUtil.GetDigitCount` - Returns the number of digits in an int without needing to convert to string.
 - Add unit tests for `MathUtil`
    - For the prime number test I added some reasonable testing that should hit all code paths but checking that the first `int.MaxValue` prime numbers are correct is just making unreasonable work for the unit tests!

### What is the current behaviour?
`GetReadableName`
 -  Does not include the names of outer types on a given type.
 - Would generate an incorrect name when an inner type had a generic outer type
    - Ex: `MyType<int>.MyInnerType` would resolve to `MyInnerTy<int>`

### What is the new behaviour?
 - Outer types are included in the generated name
 - All combinations of generic and inner types are handled correctly
    - `MyType<int>.MyInnerType`
    - `MyType<int>.MyInnerType<float>`
    - `MyType.MyInnerType<float>`

### What issues does this resolve?
None

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [x] Yes
 - [ ] No

Migration instructions:
 - Any use of `TypeExtension.GetReadableName()` must now add `using Anvil.CSharp.Logging`
 - Any use of `MathUtil.GetDigitCount()` must now add `using Anvil.CSharp.Logging` and convert the use to an extension method.
    - Ex: `MathUtil.GetDigitCount(13)` is now `13.GetDigitCount()`

Breaking changes in other Anvil repos that depend on this one:
 - https://github.com/decline-cookies/anvil-unity-dots/pull/96

### CODE REVIEW

File: Mathematics/MathUtil.cs [hunk 1]
```diff
@@ -43,5 +43,44 @@ public static long FindPrimeNumber(int nthPrimeNumberToFind)
...
+                if (value < 100) return 2;
+                if (value < 1_000) return 3;
+                if (value < 10_000) return 4;
+                if (value < 100_000) return 5;
+                if (value < 1_000_000) return 6;
+                if (value < 10_000_000) return 7;
+                if (value < 100_000_000) return 8;
+                if (value < 1_000_000_000) return 9;
+                return 10;
+            }
```
Comment (tjvezina) [inline Mathematics/MathUtil.cs:69]: Could you just use `Math.Log10`?
```cs
return Math.Floor(Math.Log10(Math.Abs(value))) + 1
```
or is this far more performant or something?

File: Reflection/TypeExtension.cs [hunk 2]
```diff
@@ -48,24 +51,98 @@ public static bool isStatic(this Type type)
...
+        /// <param name="type">The type to get a readable name for.</param>
+        /// <param name="genericArgs">
+        /// The generic arguments of the type ordered from outer to inner generic arguments.
+        /// Ex: MyType<genericArgs[0]>.MyInnerType<genericArgs[1]>
+        /// (The default order returned by <see cref="Type.GenericTypeArguments"/>
+        /// </param>
+        /// <param name="genericArgsUpperBound">
+        /// The upper bounds (exclusive) of the <see cref="genericArgs"/> array to evaluate
+        /// </param>
+        /// <returns></returns>
```
Comment (tjvezina) [inline Reflection/TypeExtension.cs:82]: Blank return value description

File: Reflection/TypeExtension.cs [hunk 3]
```diff
@@ -48,24 +51,98 @@ public static bool isStatic(this Type type)
...
             }
 
             if (type.GetGenericTypeDefinition() == s_NullableType)
             {
-                return $"{type.GenericTypeArguments[0].GetReadableName()}?";
+                return $"{outerTypes}{genericArgs[genericArgsLowerBound].GetReadableName()}?";
             }
 
             // Remove the generic type count indicator (`n) from the type name
             // Avoid having to calculate the number of digits in the generic type count by assuming it's under 100
```
Comment (tjvezina) [inline Reflection/TypeExtension.cs:116]: The second line of this comment can be removed now, as it's no longer using a shortcut

**Review by tjvezina — COMMENTED** [review]
Thanks for fixing that! Just a couple of comments and a suggestion.

File: Mathematics/MathUtil.cs (hunk 1 above)
Comment (mbaker3) [reply Mathematics/MathUtil.cs:69]: Supposedly this is faster...well at least faster the smaller the number is....which seems like the most likely use case for this.

The SO post I linked in the doc comments the guy went through some tests. I haven't taken the time to validate the performance myself.
https://stackoverflow.com/a/51099524/640196

File: Reflection/TypeExtension.cs (hunk 3 above)
Comment (mbaker3) [reply Reflection/TypeExtension.cs]: Thanks, fixed!

File: Reflection/TypeExtension.cs (hunk 2 above)
Comment (mbaker3) [reply Reflection/TypeExtension.cs]: Good catch. Fixed!

**Review by tjvezina — DISMISSED** [review]
The implementation of `GetDigitCount` makes me scowl, but it's certainly accurate.

**Review by jkeon — DISMISSED** [review]

**Comment by mbaker3** [comment]
Re-requesting review. After discussing I ended up moving all of the readability based utility methods into `ReadabilityExtension` and updating the tests as needed. Functionally this is the same but does introduce a breaking change in `anvil-unity-dots` which is resolved by this PR: https://github.com/decline-cookies/anvil-unity-dots/pull/96

File: Tests/Logging/ReadabilityExtensionTests.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+            Assert.That(typeof(TestOuterGenericNestedClass<>.TestInnerGenericClass<>).GetReadableName(),
+                Is.EqualTo("ReadabilityExtensionTests.TestOuterGenericNestedClass<>.TestInnerGenericClass<>"));
+        }
+
+        [Test]
+        public static void GetDigitCount()
+        {
+            Assert.That(0.GetDigitCount(), Is.EqualTo(1));
+            Assert.That(0.GetDigitCount(true), Is.EqualTo(1));
```
Comment (jkeon) [inline Tests/Logging/ReadabilityExtensionTests.cs:82]: Neat, i didn't know you could call methods on a literal.

**Review by jkeon — APPROVED** [review]

File: Tests/Reflection/TypeExtensionTests.cs [hunk 5]
```diff
@@ -1,32 +1,11 @@
-using NUnit.Framework;
-using Anvil.CSharp.Reflection;
+
 
 namespace Anvil.CSharp.Tests
 {
     public static class TypeExtensionTests
     {
```
Comment (tjvezina) [inline Tests/Reflection/TypeExtensionTests.cs:6]: Is this just an empty file now? Can it be removed?

Comment (mbaker3) [reply Tests/Reflection/TypeExtensionTests.cs:6]: #131 Incoming!

File: Tests/Logging/ReadabilityExtensionTests.cs (hunk 4 above)
Comment (mbaker3) [reply Tests/Logging/ReadabilityExtensionTests.cs:82]: Yea, you do have to be careful about expressions like this though.
`-1.GetDigitCount(true)`. The `-` gets applied after the method call.

The fix is
`(-1).GetDigitCount(true)`

**Review by jkeon — APPROVED** [review]

**Pull request merged**